pkg/apiserver/rest/dualwriter_mode1.go

```diff
@@ -27,36 +33,193 @@ func NewDualWriterMode1(legacy LegacyStorage, storage Storage) *DualWriterMode1
+	log := d.Log.WithValues("name", name, "kind", options.Kind)
 	ctx = klog.NewContext(ctx, d.Log)
-	return d.Legacy.Delete(ctx, name, deleteValidation, options)
+	var method = "delete"
+
+	startLegacy := time.Now().UTC()
+	res, async, err := d.Legacy.Delete(ctx, name, deleteValidation, options)
+	if err != nil {
+		log.Error(err, "unable to delete object in legacy storage")
+		d.recordLegacyDuration(true, mode, name, method, startLegacy)
+		return res, async, err
+	}
+	d.recordLegacyDuration(false, mode, name, method, startLegacy)
+
+	go func() {
+		startStorage := time.Now().UTC()
+		ctx, _ := context.WithTimeoutCause(ctx, time.Second*10, errors.New("storage delete timeout"))
+		_, _, err := d.Storage.Delete(ctx, name, deleteValidation, options)
+		defer d.recordStorageDuration(err != nil, mode, name, method, startStorage)
+	}()
+
+	return res, async, err
 }
 
 // DeleteCollection overrides the behavior of the generic DualWriter and deletes only from LegacyStorage.
 func (d *DualWriterMode1) DeleteCollection(ctx context.Context, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions, listOptions *metainternalversion.ListOptions) (runtime.Object, error) {
-	ctx = klog.NewContext(ctx, d.Log)
-	return d.Legacy.DeleteCollection(ctx, deleteValidation, options, listOptions)
+	log := d.Log.WithValues("kind", options.Kind, "resourceVersion", listOptions.ResourceVersion)
+	ctx = klog.NewContext(ctx, log)
+	var method = "delete-collection"
+
+	startLegacy := time.Now().UTC()
+	res, err := d.Legacy.DeleteCollection(ctx, deleteValidation, options, listOptions)
+	if err != nil {
+		log.Error(err, "unable to delete collection in legacy storage")
+		d.recordLegacyDuration(true, mode, options.Kind, method, startLegacy)
+	}
+	d.recordLegacyDuration(false, mode, options.Kind, method, startLegacy)
+
+	go func() {
+		startStorage := time.Now().UTC()
+		ctx, _ := context.WithTimeoutCause(ctx, time.Second*10, errors.New("storage deletecollection timeout"))
+		_, err := d.Storage.DeleteCollection(ctx, deleteValidation, options, listOptions)
+		defer d.recordStorageDuration(err != nil, mode, options.Kind, method, startStorage)
+	}()
+
+	return res, err
 }
 
 func (d *DualWriterMode1) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
-	ctx = klog.NewContext(ctx, d.Log)
-	return d.Legacy.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
+	log := d.Log.WithValues("name", name, "kind", options.Kind)
+	ctx = klog.NewContext(ctx, log)
+	var method = "update"
+
+	startLegacy := time.Now().UTC()
+	res, async, errLegacy := d.Legacy.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
+	if errLegacy != nil {
```

Comment:
don't we want to return if this fails?

Reply:
so, we are returning the error in [here](https://github.com/grafana/grafana/pull/87739/files#diff-86a0e03f6dfd59cc03a1c164ac9565e99186977647fce8b4563529605a25b7e0R222).

My thinking on this was, we don't return the error immediately, so we have a chance to write / read the object in the object store, so that we can better measure the success rate.
In this way, the principle of mode1 can be "no store will block the other", whether in mode2 it can be "failure in the first store we write / read affects dual writer success rate"

Reply:
I tend to agree that if the legacy update fails we shouldn't attempt the k8s update at all, since otherwise we know we are causing drift between the 2 stores.

Reply:
I would generally agree with Dan. I would, however, be curious to know if a failing write to legacy succeeds in unified storage. That said, I would be ok to have both storages drift and see why legacy failed while unified storage succeeded, maybe that would be a condition we need to fix in unified storage. Plus, we don't care to have drifting storages in mode 1.

Reply:
I would actually disagree because this is mode1. Initially, it wasn't even supposed to be dealing with object storage in the first place. We decided to change this so we could better data in regards to latency and availability from an earlier stage. In order to do that, I think we need to give both stores the same fair chance of success.
Moreover, we will be working on a syncer soon, which hopefully gives us the ability to better sync between the 2.
I think having this data as soon as possible and understanding where we are drifting to both ways, could give us more room to fix things while we still have small data.
Finally, we will not be staying in mode1 for long (I hope 😅 ) since mode2 is what we want to aim for.
And in mode2, we already have the behaviour of returning an error if either call fails.
Wdyt?

Reply:
> In order to do that, I think we need to give both stores the same fair chance of success.

I'm not sure I understand the reasoning here, if we look at the stats we can still easily compute success rates and have a fair comparison without trying to write things to unified storage that we already know will never appear in the legacy store.  We'll also run into issues during sync if the entry in unified storage appears to be newer than the entry in the legacy store.

Reply:
hm, that's a good point, we can reduce the syncer's complexity to only sync one direction towards object store... I'll return an error for all the write methods then (`create`, `update`, `delete` and `deletecollection`)